app/grandchallenge/archives/models.py

```diff
@@ -13,4 +17,42 @@ class Archive(UUIDModel):
     images = models.ManyToManyField(Image)
 
     def __str__(self):
-        return "<{} {}>".format(self.__class__.__name__, self.name)
+        return f"<{self.__class__.__name__} {self.name}>"
+
+    def delete(self, *args, **kwargs):
+        """
+        Removes all patients, studies, images, imagefiles and annotations that belong
+        exclusively to this archive
+        """
+        images_to_remove = (
+            Image.objects.annotate(num_archives=Count("archive"))
+            .filter(archive__id=self.id, num_archives=1)
+            .all()
+        )
+        protected_study_ids = set()
+        protected_patient_ids = set()
+        for image in images_to_remove:
+            if image.study is None:
+                continue
+            for image1 in image.study.image_set.all():
+                if image1 not in images_to_remove:
+                    protected_study_ids.add(image.study.id)
+                    protected_patient_ids.add(image.study.patient.id)
+                    break
+
+        related_patients = Patient.objects.filter(
+            study__image__in=images_to_remove
+        ).distinct()
+        related_studies = Study.objects.filter(
+            image__in=images_to_remove
+        ).distinct()
+        for patient in related_patients:
+            if patient.id not in protected_patient_ids:
+                patient.delete(*args, **kwargs)
+        for study in related_studies:
```

Comment:
I think these 2 for loops can be replaced with an extra filter in each of the queries, eg `.exclude(pk__in=protected_patient_ids)`